src/harness/sas_objects.py

```diff
@@ -0,0 +1,387 @@
+            grant_requests: A list of dictionary with a single key-value pair
+                where the key is"grantRequest" and the value is a list of
+                individual CBSD grant requests (each of which is itself a dictionary).
+                There should be exactly one grant request per registration request.
+        """
+        #Checking if the number of registration requests matches number of grant requests
+        #There should be exactly one grant request per registration request.
+        assert (len(grant_requests) == len(registration_requests))
+        cbsd_ids = self.testcase.assertRegistered(registration_requests)
+
+        #Make one CBSD object per successful grant request
+        for index, grant_request in enumerate(grant_requests):
+            grant_request["grantRequest"][0]["cbsdId"] = cbsd_ids[index]
+            grant_response = self.testcase._sas.Grant(grant_request, self.ssl_cert, self.ssl_key)[
+                'grantResponse'][0]
+            if grant_response['response']['responseCode'] == 0:
+                cbsdobject = Cbsd(cbsd_ids[index], registration_requests[index],
+                                  [grant_response['grantId']], [grant_request])
+                self.cbsd_objects.update(
+                    {grant_response['cbsdId']: cbsdobject})
+
+    def getCbsdObjectById(self, cbsd_id):
+        return self.cbsd_objects[cbsd_id]
+
+    def heartbeatRequestForAllGrants(self):
+        """
+            This function performs heartbeat request for all grants
+        """
+        for cbsd_object_item in self.cbsd_objects.values():
+            heartbeat_requests = cbsd_object_item.constructHeartbeatRequestForAllActiveGrants()
+            for heartbeat_request in heartbeat_requests:
+                heartbeat_response = self.testcase._sas.Heartbeat(
+                    heartbeat_request, self.ssl_cert, self.ssl_key)['heartbeatResponse'][0]
+                grant_object = cbsd_object_item.grant_objects[heartbeat_response['grantId']]
+                if heartbeat_response['response']['responseCode'] == 0:
+                    transmit_expire_time = datetime.strptime(heartbeat_response['transmitExpireTime'],
+                                                             '%Y-%m-%dT%H:%M:%SZ')
+                    self.testcase.assertLess(datetime.utcnow(), transmit_expire_time)
+                    self.testcase.assertLessEqual(
+                        (transmit_expire_time - datetime.utcnow()).total_seconds(), 240)
+                    grant_object.setState(Grant.AUTHORIZED)
+                else:
+                    self.mapResponseCodeToGrantState(
+                        heartbeat_response['response']['responseCode'], grant_object)
+
+    def mapResponseCodeToGrantState(self, heartbeat_response_code, grant_object):
+        """
+            This function maps the heartbeat response code to the grant state
+        """
+        if (heartbeat_response_code == 502):
```

Comment:
The behavior should be as follows:

0: authorized_in_last_heartbeat -> True
500: is_terminated -> True, authorized_in_last_heartbeat -> False
501: authorized_in_last_heartbeat -> False
502: relinquish Grant (send a request to the SAS), is_terminated -> True

Reply:
Updated   as per suggestion. The success case (0)    and the 502 case to relinquish the grant is handled in the calling   function other values updated as per suggestion.